test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go

```diff
@@ -0,0 +1,200 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"golang.org/x/oauth2/google"
+	clientset "k8s.io/client-go/kubernetes"
+
+	. "github.com/onsi/ginkgo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+
+	gcm "google.golang.org/api/monitoring/v3"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/discovery"
+	kubeaggrcs "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+	customclient "k8s.io/metrics/pkg/client/custom_metrics"
+)
+
+var _ = instrumentation.SIGDescribe("Stackdriver Monitoring", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+	})
+
+	f := framework.NewDefaultFramework("stackdriver-monitoring")
+	var kubeClient clientset.Interface
+	var kubeAggrClient kubeaggrcs.Interface
+	var customMetricsClient customclient.CustomMetricsClient
+	var discoveryClient *discovery.DiscoveryClient
+
+	It("should run Custom Metrics - Stackdriver Adapter [Feature:StackdriverMonitoring]", func() {
+		kubeClient = f.ClientSet
+		kubeAggrClient = f.AggregatorClient
+		config, err := framework.LoadConfig()
+		if err != nil {
+			framework.Failf("Failed to load config: %s", err)
+		}
+		customMetricsClient = customclient.NewForConfigOrDie(config)
+		discoveryClient = discovery.NewDiscoveryClientForConfigOrDie(config)
+		testAdapter(f, kubeClient, customMetricsClient, discoveryClient)
+	})
+})
+
+func testAdapter(f *framework.Framework, kubeClient clientset.Interface, customMetricsClient customclient.CustomMetricsClient, discoveryClient *discovery.DiscoveryClient) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, gcm.CloudPlatformScope)
+
+	// Hack for running tests locally
+	// If this is your use case, create application default credentials:
+	// $ gcloud auth application-default login
+	// and uncomment following lines (comment out the two lines above):
+	/*
+		ts, err := google.DefaultTokenSource(oauth2.NoContext)
+		framework.Logf("Couldn't get application default credentials, %v", err)
+		if err != nil {
+			framework.Failf("Error accessing application default credentials, %v", err)
+		}
+		client := oauth2.NewClient(oauth2.NoContext, ts)
+	*/
+
+	gcmService, err := gcm.New(client)
+	if err != nil {
+		framework.Failf("Failed to create gcm service, %v", err)
+	}
+
+	framework.ExpectNoError(err)
+
+	// Set up a cluster: create a custom metric and set up k8s-sd adapter
+	err = createDescriptors(gcmService, projectId)
```

Comment:
I think that with the current implementation you will never cleanup descriptors if you leak for whatever reason the metric. Would it make sense to delete metric from in case it already exists? If you agree I'm fine with leaving it as a TODO and fixing it in a follow-up PR.

Reply:
Added TODO. Not sure if I understood your comment correctly, but in general I agree that we should clean up time series as well.